include/exadg/time_integration/am_constants.cpp

```diff
@@ -0,0 +1,159 @@
+ *
+ *  Copyright (C) 2021 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+// deal.II
+#include <deal.II/base/exceptions.h>
+
+// ExaDG
+#include <exadg/time_integration/am_constants.h>
+
+namespace ExaDG
+{
+AMTimeIntegratorConstants::AMTimeIntegratorConstants(unsigned int const order,
+                                                     bool const         start_with_low_order)
+  : TimeIntegratorConstantsBase(order, start_with_low_order), gamma0(-1.0), alpha(order - 1)
+{
+  AssertThrow(order >= 1 && order <= 4,
+              dealii::ExcMessage("Specified order of Adams-Moulton scheme not implemented."));
+
+  // The default case is start_with_low_order = false.
+  set_constant_time_step(order);
+}
+
+
+double
+AMTimeIntegratorConstants::get_gamma0() const
+{
+  AssertThrow(gamma0 > 0.0, dealii::ExcMessage("Constant gamma0 has not been initialized."));
+
+  return gamma0;
+}
+
+double
+AMTimeIntegratorConstants::get_alpha(unsigned int const i) const
+{
+  AssertThrow(i < alpha.size(),
+              dealii::ExcMessage(
+                "In order to access time integrator constants, the index "
+                "has to be smaller than the order of the time integration scheme."));
+
+  return alpha[i];
+}
+
+void
+AMTimeIntegratorConstants::print(dealii::ConditionalOStream & pcout) const
+{
+  pcout << "Gamma0 = " << gamma0 << std::endl;
+  for(unsigned int i = 0; i < alpha.size(); ++i)
+    pcout << "Alpha[" << i << "] = " << alpha[i] << std::endl;
+}
+
+void
+AMTimeIntegratorConstants::set_constant_time_step(unsigned int const current_order)
+{
+  AssertThrow(current_order <= order,
+              dealii::ExcMessage(
+                "There is a logical error when updating the AM time integrator constants."));
+
+  if(current_order == 1) // AM 1
+  {
+    gamma0 = 1.0;
+  }
+  else if(current_order == 2) // AM 2
+  {
+    gamma0   = 1.0 / 2.0;
+    alpha[0] = 1.0 / 2.0;
+  }
+  else if(current_order == 3) // AM 3
+  {
+    gamma0   = 5.0 / 12.0;
+    alpha[0] = 8.0 / 12.0;
+    alpha[1] = -1.0 / 12.0;
+  }
+  else if(current_order == 4) // AM 4
+  {
+    gamma0   = 9.0 / 24.0;
+    alpha[0] = 19.0 / 24.0;
+    alpha[1] = -5.0 / 24.0;
+    alpha[2] = 1.0 / 24.0;
+  }
```

Comment:
This way of implementation is not very nice I think (which I probably introduced myself ...). I think the motivation was to write the assert only once and not in several functions (like the `set_...()` functions). I think a `switch` with `default` branch (calling a private member function `assert_order()`) would be better.

I make this comment here as a reminder for future work. No need to address this in the present PR.

Reply:
I would also be fine to change it within the PR. Just let me know which way you prefer :)

Reply:
I am certainly happy if you want to care about such a refactoring!

Reply:
[ac5015c](https://github.com/exadg/exadg/pull/469/commits/ac5015c1a7c94f70ddf90a8d4c9a9bbc7f72acf0)

Reply:
I don't think we need `assert_order()`. Arriving at the default `assert_order()` would always trigger. I just placed a standard assert in the default branch.